test/AutoDiff/differentiable_func_debuginfo.swift

```diff
@@ -0,0 +1,37 @@
+// RUN: %target-swift-frontend -c %s -g -O -parse-stdlib -parse-as-library -module-name Swift
+
+// TF-597: Exact minimal reproducer for IRGenDebugInfo crash.
+// Crash occurred only with `-g` and `-O`.
+//
+// ```
+// Assertion failed: (OffsetInBits + SizeInBits <= getSizeInBits(Var) && "pars > totum"),
+// function emitVariableDeclaration, file swift/lib/IRGen/IRGenDebugInfo.cpp, line 2216.
+// Stack dump:
+// 1.	Swift version 5.1-dev (LLVM 200186e28b, Swift c09c14dec5)
+// 2.	While emitting IR SIL function "@$ss8pullback2at2inyx_q_xXFts14DifferentiableRzsADR_r0_lF".
+//  for 'pullback(at:in:)' (at swift/test/AutoDiff/differentiable_func_debuginfo.swift:21:8)
+// ```
+//
+// The crash was because `IRGenDebugInfoImpl::getOrCreateType` computes
+// `llvm::DIType` type debug info by demangling type names.
+//
+// Since `@differentiable` and `@differentiable(linear)` function types did
+// not have mangling support, `getOrCreateType` computed a regular `(A) -> B`
+// function type instead of a `@differentiable (A) -> B` function type, leading
+// to a type size inconsistency.
+//
+// Conclusion: mangling coverage is important.
+
+// Minimal dummy compiler-known `Differentiable` protocol.
+public protocol Differentiable {
+  associatedtype TangentVector
+}
+
+// This declaration is necessary to reproduce the crash for some reason.
```

Comment:
Could you please update this comment?

What is happening here is pretty simple.  `Builtin.autodiffApply_jvp` is constructing a use of the `tf597ProblematicVarDecl` type which would be decorated without the `@differential`.  The parameter for `blackHole` is of type `$@noescape @callee_guaranteed (@in_guaranteed T) -> @out U` which will match the decorated name for the differential function when the `@differential` attribute is not accounted for.  As a result, when generating the debug information the types are uniqued the two values are treated as the same except the type in `blackHole` is smaller than the `@differential` type which causes the validation to fail.

Reply:
That makes sense, thanks!

Reply:
Done in https://github.com/apple/swift/pull/26600.